metrics/otelgrpc/otelgrpc.go

```diff
@@ -0,0 +1,198 @@
+package otelgrpc
+
+import (
+	"context"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/goto/meteor/utils"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	semconv "go.opentelemetry.io/otel/semconv/v1.20.0"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/protobuf/proto"
+)
+
+type UnaryParams struct {
+	Start  time.Time
+	Method string
+	Req    any
+	Res    any
+	Err    error
+}
+
+type Monitor struct {
+	duration     metric.Int64Histogram
+	requestSize  metric.Int64Histogram
+	responseSize metric.Int64Histogram
+	attributes   []attribute.KeyValue
+}
+
+func NewOtelGRPCMonitor(hostName string) Monitor {
+	meter := otel.Meter("github.com/goto/meteor/metrics/otelgrpc")
+
+	duration, err := meter.Int64Histogram("rpc.client.duration", metric.WithUnit("ms"))
+	handleOtelErr(err)
+
+	requestSize, err := meter.Int64Histogram("rpc.client.request.size", metric.WithUnit("By"))
+	handleOtelErr(err)
+
+	responseSize, err := meter.Int64Histogram("rpc.client.response.size", metric.WithUnit("By"))
+	handleOtelErr(err)
+
+	addr, port := ExtractAddress(hostName)
+
+	return Monitor{
+		duration:     duration,
+		requestSize:  requestSize,
+		responseSize: responseSize,
+		attributes: []attribute.KeyValue{
+			semconv.RPCSystemGRPC,
+			attribute.String("network.transport", "tcp"),
+			attribute.String("server.address", addr),
+			attribute.String("server.port", port),
+		},
+	}
+}
+
+func GetProtoSize(p any) int {
+	if p == nil {
+		return 0
+	}
+
+	size := proto.Size(p.(proto.Message))
+	return size
+}
+
+func (m *Monitor) RecordUnary(ctx context.Context, p UnaryParams) {
+	reqSize := GetProtoSize(p.Req)
+	resSize := GetProtoSize(p.Res)
+
+	var attrs []attribute.KeyValue
+	copy(attrs, m.attributes)
+	attrs = append(attrs, attribute.String("rpc.grpc.status_text", utils.StatusText(p.Err)))
+	attrs = append(attrs, attribute.String("network.type", netTypeFromCtx(ctx)))
+	attrs = append(attrs, ParseFullMethod(p.Method)...)
+
+	m.duration.Record(ctx,
+		time.Since(p.Start).Milliseconds(),
+		metric.WithAttributes(attrs...))
+
+	m.requestSize.Record(ctx,
+		int64(reqSize),
+		metric.WithAttributes(attrs...))
+
+	m.responseSize.Record(ctx,
+		int64(resSize),
+		metric.WithAttributes(attrs...))
+}
+
+func (m *Monitor) RecordStream(ctx context.Context, start time.Time, method string, err error) {
+	var attrs []attribute.KeyValue
```

Comment:
needs to be allocated first here as well. Or alternatively, you can use append:

```go
attrs := append([]attribute.KeyValue(nil), m.attributes...)
```

Reply:
using make method for this